### PR TITLE
DataMapper support with Doctrine tests

### PR DESCRIPTION
**Pull request by DavertMik**
Now you can use have FactoryMuffin to work with DataMapper pattern, especially with Doctrine. 
To have FactoryMuffin to be initialized with Doctine you should use `RepositoryStore` instead of `ModelStore` class:

```php
$em = EntityManager::create($dbParams, $config);
$fm = new FactoryMuffin(new RepositoryStore($em));
```

Unlike model store you can't use `factory` prefix to create associated models. `factory` creation method is supported to create a record and recturn its id, while Doctrine expects to receive a complete entity into association. That's why you should use `entity` to get a record into association

```php
'user_id' => 'factory|User', // to receive id of a record
// vs
'user' => 'entity|User', // to receive a record
```

Also Doctrine added as dev dependency and tests for it added.

### CODE REVIEW

File: src/RepositoryStore.php [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            throw new DeleteMethodNotFoundException(get_class($this->storage), $method);
+        }
+
+        $this->storage->$method($model);
+
+        return true;
+    }
+
+    /**
+     * {@inheritdoc}
```
Comment (GrahamCampbell) [inline src/RepositoryStore.php]: Please don't use this tag.

File: src/RepositoryStore.php [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+     *
+     * @throws DeletingFailedException
```
Comment (GrahamCampbell) [inline src/RepositoryStore.php]: Please reference FCQNs in phpdoc always.

File: src/RepositoryStore.php [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+     */
+    public function deleteSaved()
+    {
+        parent::deleteSaved();
```
Comment (GrahamCampbell) [inline src/RepositoryStore.php]: Please add a newline after this.

File: tests/AbstractTestCase.php [hunk 4]
```diff
@@ -32,7 +35,10 @@ public static function setupBeforeClass()
 
     public static function tearDownAfterClass()
     {
-        static::$fm->deleteSaved();
+        try {
+            static::$fm->deleteSaved();
+        } catch (Exception $e) {
```
Comment (GrahamCampbell) [inline tests/AbstractTestCase.php]: ?

File: tests/DoctrineTest.php [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+/**
+ * This is eloquent test class.
+ *
+ * @author Graham Campbell <graham@alt-three.com>
+ */
+class DoctrineTest extends AbstractTestCase
+{
+    const USER_ENTITY = 'League\FactoryMuffin\Test\User';
+    const CAT_ENTITY = 'League\FactoryMuffin\Test\Cat';
+    /**
```
Comment (GrahamCampbell) [inline tests/DoctrineTest.php]: missing newline

**Comment by GrahamCampbell** [comment]
Thanks. I've left a few comments.

**Comment by GrahamCampbell** [comment]
Could you skip the failing tests on hhvm too please?

File: tests/AbstractTestCase.php (hunk 4 above)
Comment (DavertMik) [inline tests/AbstractTestCase.php]: oups. debug stuff

**Comment by DavertMik** [comment]
Done.

File: src/Generators/EntityGenerator.php [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+     */
+    private $factoryMuffin;
+
+    /**
+     * Create a new factory generator instance.
+     *
+     * @param string                              $kind          The kind of attribute.
+     * @param object                              $model         The model instance.
+     * @param \League\FactoryMuffin\FactoryMuffin $factoryMuffin The factory muffin instance.
+     */
```
Comment (GrahamCampbell) [inline src/Generators/EntityGenerator.php]: missing return void

File: src/RepositoryStore.php [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+        return true;
+    }
+
+    /**
+     * Flushes changes to storage.
+     *
+     * @throws MethodNotFoundException
+     *
+     * @return mixed
```
Comment (GrahamCampbell) [inline src/RepositoryStore.php]: ?

File: src/RepositoryStore.php [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+     */
+    protected function flush()
+    {
+        $method = $this->flushMethod;
+
+        if (!method_exists($this->storage, $method)) {
+            throw new MethodNotFoundException(get_class($this->storage), $method, "Can't save: flush method not found");
+        }
+        $this->storage->$method();
```
Comment (GrahamCampbell) [inline src/RepositoryStore.php]: missing newline before this

File: src/RepositoryStore.php [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+     * @return mixed
+     */
+    protected function save($model)
+    {
+        $method = $this->saveMethod;
+
+        if (!method_exists($this->storage, $method)) {
+            throw new SaveMethodNotFoundException(get_class($this->storage), $method);
+        }
+        $this->storage->$method($model);
```
Comment (GrahamCampbell) [inline src/RepositoryStore.php]: same

File: src/RepositoryStore.php [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+
+    /**
+     * Save our object to the db, and keep track of it.
+     *
+     * @param object $model The model instance.
+     *
+     * @throws \League\FactoryMuffin\Exceptions\SaveMethodNotFoundException
+     * @throws \League\FactoryMuffin\Exceptions\MethodNotFoundException
+     *
+     * @return mixed
```
Comment (GrahamCampbell) [inline src/RepositoryStore.php]: always returns true as far as i can see

Comment (GrahamCampbell) [inline src/RepositoryStore.php]: maybe returning `1` would be more consistent though?

File: tests/DoctrineTest.php [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+     */
+    protected static $em;
+
+    public static function setupBeforeClass()
+    {
+        $dbParams = [
+            'driver'   => 'pdo_sqlite',
+            'memory'   => true,
+        ];
+        $entitiesPath = [__DIR__.'/entities'];
```
Comment (GrahamCampbell) [inline tests/DoctrineTest.php]: doesn't need defining early.

File: src/RepositoryStore.php (hunk 7 above)
Comment (DavertMik) [inline src/RepositoryStore.php]: that what PHPStorm adds

File: src/Generators/EntityGenerator.php (hunk 6 above)
Comment (DavertMik) [inline src/Generators/EntityGenerator.php]: strange, PHPStorm marks `@return void` in constructor as error and updates it
I don't feel it is necessary

File: src/RepositoryStore.php (hunk 10 above)
Comment (DavertMik) [inline src/RepositoryStore.php]: on other side, I can catch any exception and return false if exception is thrown...

File: tests/DoctrineTest.php (hunk 11 above)
Comment (DavertMik) [inline tests/DoctrineTest.php]: sorry, I didn't get this point

**Comment by DavertMik** [comment]
updtd

File: src/RepositoryStore.php (hunk 7 above)
Comment (GrahamCampbell) [inline src/RepositoryStore.php]: well, it's not correct?

File: src/Generators/FactoryGenerator.php [hunk 12]
```diff
@@ -23,29 +21,11 @@
...
-     */
-    private $model;
-
-    /**
-     * The factory muffin instance.
-     *
-     * @var \League\FactoryMuffin\FactoryMuffin
-     */
-    private $factoryMuffin;
+    const PREFIX = 'factory|';
```
Comment (GrahamCampbell) [inline src/Generators/FactoryGenerator.php]: :-1: for dealing with prefixes like this

Comment (GrahamCampbell) [inline src/Generators/FactoryGenerator.php]: maybe we should calculuate them from the class name?

Comment (DavertMik) [inline src/Generators/FactoryGenerator.php]: > :-1: for dealing with prefixes like this

so you think that hardcoding 8 was better? https://github.com/thephpleague/factory-muffin/blob/master/src%2FGenerators%2FFactoryGenerator.php#L89

I don't think that prefix should be bound to class name.  At least CallableGenerator does not have any pefixes, so we can't stick to its name.

Comment (GrahamCampbell) [inline src/Generators/FactoryGenerator.php]: No. In this case, we should implement a method to get the prefix and have that on the interface.

Comment (DavertMik) [inline src/Generators/FactoryGenerator.php]: ok

**Comment by DavertMik** [comment]
done

**Comment by DavertMik** [comment]
Ping-ping?

File: src/RepositoryStore.php [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+     * Methods to flush changes to storage.
+     *
+     * @var string
+     */
+    protected $flushMethod = 'flush';
+
+    /**
+     * Instance of a class that performs flushes (EntityManager for Doctrine).
+     *
+     * @var
```
Comment (GrahamCampbell) [inline src/RepositoryStore.php]: ?

File: src/RepositoryStore.php [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+        $this->storage->$method($model);
+        $this->flush();
+
+        return true;
+    }
+
+    /**
+     * Flushes changes to storage.
+     *
+     * @throws MethodNotFoundException
```
Comment (GrahamCampbell) [inline src/RepositoryStore.php]: please use yjr fqcn

**Comment by DavertMik** [comment]
Done. But I hope that was the last part of codestyling and dockblock bureaucracy )

**Comment by GrahamCampbell** [comment]
Thanks for all your work here. I appreciate it very much. I'll do a final review of this at the weekend and then merge. :)

**Comment by DavertMik** [comment]
Ok, will be waiting for that and for final release of 3.0 :+1:

**Comment by DavertMik** [comment]
@GrahamCampbell yep, about 3.0. With this PR I have plans to include factory-muffin in Codeception 2.2 but I need to know your release plans so to decide If I can stick to it. It's not a nice idea to rely on beta dependency (

**Comment by GrahamCampbell** [comment]
Thank you @DavertMik.

File: src/RepositoryStore.php [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+     * @throws \League\FactoryMuffin\Exceptions\DeletingFailedException
+     *
+     * @return void
+     */
+    public function deleteSaved()
+    {
+        parent::deleteSaved();
+
+        try {
+            $this->storage->flush();
```
Comment (GrahamCampbell) [inline src/RepositoryStore.php]: @DavertMik I assume this was meant to be `$this->flush()`, otherwise we're ignoring the flushMethod property?

**Comment by GrahamCampbell** [comment]
@DavertMik v3.0.0 is released now, even with incomplete upgrading docs. That was the main reason I was holding off, but it'a already like 18 months late, so I'm just going to ship it.

**Comment by GrahamCampbell** [comment]
Factory Muffin 3.x will be maintained into the future, and follows semver.

**Comment by DavertMik** [comment]
Awesome!

**Pull request closed without merging**